src/plugins/intel_gpu/include/intel_gpu/primitives/data.hpp

```diff
@@ -4,37 +4,195 @@
+        this->bin_offset = bin_offset;
+        this->original_size = original_size;
+        this->original_dtype = original_dtype;
+        this->curr_dtype = curr_dtype;
+        do_weightless_caching = true;
+
+        if (original_dtype != curr_dtype) {
+            do_precision_conversion = true;
+        }
+    }
+
+    void invalidate() {
+        do_weightless_caching = false;
+    }
+
+    void set_new_dtype(ov::element::Type curr_dtype) {
+        this->curr_dtype = curr_dtype;
+        do_precision_conversion = original_dtype != curr_dtype;
+    }
+
+    bool save(BinaryOutputBuffer& ob, size_t data_size) const {
+        if (!do_weightless_caching) {
+            ob << false;
+            return false;
+        }
+
+        ob << true;
+        ob << bin_offset;
+        ob << do_precision_conversion;
+        if (do_precision_conversion) {
+            ob << original_size;
+            ob << make_data(&original_dtype, sizeof(ov::element::Type));
+            ob << make_data(&curr_dtype, sizeof(ov::element::Type));
+        }
+        return true;
+    }
+
+    std::shared_ptr<weights_mem> load(BinaryInputBuffer& ib,
+                                      std::shared_ptr<ov::MappedMemory> mapped_weights,
+                                      size_t data_size) {
+        ib >> do_weightless_caching;
+        if (!do_weightless_caching) {
+            return nullptr;
+        }
+
+        OPENVINO_ASSERT(mapped_weights != nullptr, "mmap object is null");
+
+        ib >> bin_offset;
+        ib >> do_precision_conversion;
+        if (do_precision_conversion) {
+            ib >> original_size;
+            ib >> make_data(&original_dtype, sizeof(ov::element::Type));
+            ib >> make_data(&curr_dtype, sizeof(ov::element::Type));
+        } else {
+            original_size = data_size;
+        }
+
+        auto mem_obj = std::make_shared<weights_mem>();
+        mem_obj->shared_buf = std::make_shared<ov::SharedBuffer<std::shared_ptr<ov::MappedMemory>>>(
+            mapped_weights->data() + bin_offset,
+            original_size,
+            mapped_weights);
+
+        if (should_run_transformations()) {
+            run_transformations(mem_obj);
+        }
+        return mem_obj;
+    }
+
+private:
+    bool do_weightless_caching = false;
+    bool do_precision_conversion = false;
+
+    size_t bin_offset = SIZE_MAX;
+    size_t original_size = SIZE_MAX;
+    ov::element::Type original_dtype = ov::element::Type_t::undefined;
+    ov::element::Type curr_dtype = ov::element::Type_t::undefined;
+
+    bool should_run_transformations() {
+        return do_precision_conversion;
+    }
+
+    void run_transformations(std::shared_ptr<weights_mem> mem_obj) {
+        size_t num_elems = original_size / original_dtype.size();
+        ov::Shape shape({num_elems});
+        auto orig_constant =
+            std::make_shared<ov::op::v0::Constant>(original_dtype, shape, mem_obj->shared_buf);
```

Comment:
Here data are copied, form buffer to Constant, but maybe is not required?

The Constant have possibility to be created from memory pointer as shared buffer check the [ctor](https://github.com/openvinotoolkit/openvino/blob/master/src/core/include/openvino/op/constant.hpp#L198), 
it can also take `mapped_weights` as last parameter to keep this object alive if required.